File: suites/08_media_encoding_kernels/comparisons/motion_estimation.rs
fn abs_diff(a: i64, b: i64) -> i64 { if a > b { a - b } else { b - a } }

fn motion_est_checksum(frames: i64) -> i64 {
    let width: i64 = 64; let height: i64 = 64;
    let block_size: i64 = 8; let search_range: i64 = 4;
    let mut checksum: i64 = 0;
    
    for f in 0..frames {
        for by in 0..(height / block_size) {
            for bx in 0..(width / block_size) {
                let mut best_sad: i64 = 999999999;
                let mut best_dy: i64 = 0;
                let mut best_dx: i64 = 0;
                
                for dy in -search_range..=search_range {
                    for dx in -search_range..=search_range {
                        let mut sad: i64 = 0;
                        for r in 0..block_size {
                            for c in 0..block_size {
                                let cy = by * block_size + r;
                                let cx = bx * block_size + c;
                                let ry = cy + dy; let rx = cx + dx;
                                
                                let cur_pixel = ((f * width * height + cy * width + cx) * 17 + 13) % 256;
                                let mut ref_pixel: i64 = 0;
                                if ry >= 0 && ry < height && rx >= 0 && rx < width {
                                    ref_pixel = (((f - 1) * width * height + ry * width + rx) * 17 + 13) % 256;
                                }
                                sad += abs_diff(cur_pixel, ref_pixel);
                            }
                        }
                        if sad < best_sad { best_sad = sad; best_dy = dy; best_dx = dx; }
                    }
                }
                checksum = (checksum * 31 + best_sad + best_dy * 17 + best_dx * 7) % 1000000007;
            }
        }
    }
    checksum
}

fn main() { println!("{}", motion_est_checksum(8)); }

File: suites/08_media_encoding_kernels/comparisons/motion_estimation.py
def abs_diff(a: int, b: int) -> int:
    return a - b if a > b else b - a

def c_mod(val: int, m: int) -> int:
    rem = abs(val) % m
    return -rem if val < 0 else rem

def motion_est_checksum(frames: int) -> int:
    width, height, block_size, search_range = 64, 64, 8, 4
    checksum = 0
    
    for f in range(frames):
        for by in range(height // block_size):
            for bx in range(width // block_size):
                best_sad = 999999999
                best_dy, best_dx = 0, 0
                for dy in range(-search_range, search_range + 1):
                    for dx in range(-search_range, search_range + 1):
                        sad = 0
                        for r in range(block_size):
                            for c in range(block_size):
                                cy = by * block_size + r
                                cx = bx * block_size + c
                                ry, rx = cy + dy, cx + dx
                                
                                cur_pixel = c_mod((f * width * height + cy * width + cx) * 17 + 13, 256)
                                ref_pixel = 0
                                if 0 <= ry < height and 0 <= rx < width:
                                    ref_pixel = c_mod(((f - 1) * width * height + ry * width + rx) * 17 + 13, 256)
                                sad += abs_diff(cur_pixel, ref_pixel)
                        if sad < best_sad:
                            best_sad = sad
                            best_dy = dy
                            best_dx = dx
                checksum = c_mod(checksum * 31 + best_sad + best_dy * 17 + best_dx * 7, 1000000007)
    return checksum

if __name__ == "__main__":
    print(motion_est_checksum(8))
